### ISSUE TITLE
$P1R|NA

### ISSUE THREAD

**Issue by mtrussart** [RGLab/flowCore#202]
Hi,

I am having an issue when reading a FCS file. I am getting the following error where I am trying to read the file. I went up to the flowCore functions and I found out that the problem is coming from the $P1R that has been saved as NA in the preprocessing steps. 

test=read.FCS("A1.fcs", transformation = FALSE, truncate_max_range = FALSE)

Error in readFCSdata(con, offsets, txt, transformation, which.lines, scale,  : 
  $PnRNAis larger than R's numeric limit:1.79769313486232e+308
In addition: Warning messages:
1: In for (i in seq_len(n)) { :
  closing unused connection 4 (A1.fcs)
2: In readFCSdata(con, offsets, txt, transformation, which.lines, scale,  :
  NAs introduced by coercion

I went up to the flowCore functions and I found out that the problem is coming from the $P1R that has been saved as NA. This is how it looks like in files that are being read without any errors: "...P1E|0,0|$P1N|Time|$P1R|15248628|$P20B|..." and this is how it appears in this one:  "... $P1E|0,0|$P1N|Time|$P1R|NA|$P20B...". Hence the error is been raised because of the presence of NA. Here are some more details from the functions and results:

range_str <- readFCSgetPar(x, paste("$P", 1:nrpar, "R", sep=""))
range_str
   $P1R    $P2R    $P3R    $P4R    $P5R    $P6R    $P7R    $P8R    $P9R   $P10R   $P11R   $P12R 
   "NA"   "116"  "2410" "20686" "24756" "24642" "20193" "23306" "20909"  "4659" "17646" "23397" 
  $P13R   $P14R   $P15R   $P16R   $P17R   $P18R   $P19R   $P20R   $P21R   $P22R   $P23R   $P24R 
"20725" "12575" "13289"  "4375"  "1662"  "3366" "14419"  "2341" "14986" "37113"  "8871"  "2155" 
  $P25R   $P26R   $P27R   $P28R   $P29R   $P30R   $P31R   $P32R   $P33R   $P34R   $P35R   $P36R 
"15921" "29827"  "2221"  "1291"  "3242" "35051" "28849"  "1596"   "716" "17923"  "5358" "13065" 
  $P37R   $P38R   $P39R   $P40R   $P41R   $P42R   $P43R   $P44R   $P45R 
 "6921" "32316" "10013"  "3873" "31358" "40098" "44830" "45817" "59192" 

Could you please help me find out how to deal with this error? 

sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /stornext/System/data/apps/R/R-4.0.2/lib64/R/lib/libRblas.so
LAPACK: /stornext/System/data/apps/R/R-4.0.2/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
 [1] parallel  stats4    grid      stats     graphics  grDevices utils     datasets  methods  
[10] base     

other attached packages:
 [1] CATALYST_1.12.2             SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.2
 [4] DelayedArray_0.14.1         Biobase_2.48.0              GenomicRanges_1.40.0       
 [7] GenomeInfoDb_1.24.2         IRanges_2.22.2              S4Vectors_0.26.1           
[10] BiocGenerics_0.34.0         gridExtra_2.3               multcomp_1.4-13            
[13] TH.data_1.0-10              MASS_7.3-51.6               survival_3.1-12            
[16] mvtnorm_1.1-1               lme4_1.1-23                 Matrix_1.2-18              
[19] cowplot_1.0.0               Rtsne_0.15                  ConsensusClusterPlus_1.52.0
[22] FlowSOM_1.20.0              igraph_1.2.5                ComplexHeatmap_2.4.3       
[25] pheatmap_1.0.12             RColorBrewer_1.1-2          ggrepel_0.8.2              
[28] limma_3.44.3                dplyr_1.0.2                 reshape2_1.4.4             
[31] ggridges_0.5.2              ggplot2_3.3.2               matrixStats_0.56.0         
[34] flowCore_2.0.1              readxl_1.3.1               

loaded via a namespace (and not attached):
  [1] ggbeeswarm_0.6.0          minqa_1.2.4               colorspace_1.4-1         
  [4] rjson_0.2.20              ellipsis_0.3.1            rio_0.5.16               
  [7] circlize_0.4.10           cytolib_2.0.3             XVector_0.28.0           
 [10] BiocNeighbors_1.6.0       GlobalOptions_0.1.2       base64enc_0.1-3          
 [13] clue_0.3-57               rstudioapi_0.11           hexbin_1.28.1            
 [16] CytoML_2.0.5              xml2_1.3.2                codetools_0.2-16         
 [19] splines_4.0.2             emdist_0.3-1              scater_1.16.2            
 [22] jsonlite_1.7.0            nloptr_1.2.2.2            cluster_2.1.0            
 [25] png_0.1-7                 graph_1.66.0              compiler_4.0.2           
 [28] drc_3.0-1                 BiocSingular_1.4.0        tools_4.0.2              
 [31] ncdfFlow_2.34.0           rsvd_1.0.3                gtable_0.3.0             
 [34] glue_1.4.2                GenomeInfoDbData_1.2.3    flowWorkspace_4.0.6      
 [37] ggcyto_1.16.0             Rcpp_1.0.5                carData_3.0-4            
 [40] cellranger_1.1.0          vctrs_0.3.3               nlme_3.1-148             
 [43] DelayedMatrixStats_1.10.1 stringr_1.4.0             openxlsx_4.1.5           
 [46] irlba_2.3.3               lifecycle_0.2.0           gtools_3.8.2             
 [49] statmod_1.4.34            XML_3.99-0.5              zlibbioc_1.34.0          
 [52] zoo_1.8-8                 scales_1.1.1              RProtoBufLib_2.0.0       
 [55] hms_0.5.3                 RBGL_1.64.0               sandwich_2.5-1           
 [58] yaml_2.2.1                curl_4.3                  latticeExtra_0.6-29      
 [61] stringi_1.4.6             plotrix_3.7-8             BiocParallel_1.22.0      
 [64] boot_1.3-25               zip_2.1.1                 shape_1.4.4              
 [67] rlang_0.4.7               pkgconfig_2.0.3           bitops_1.0-6             
 [70] lattice_0.20-41           purrr_0.3.4               tidyselect_1.1.0         
 [73] plyr_1.8.6                magrittr_1.5              R6_2.4.1                 
 [76] nnls_1.4                  generics_0.0.2            pillar_1.4.6             
 [79] haven_2.3.1               foreign_0.8-80            withr_2.2.0              
 [82] abind_1.4-5               RCurl_1.98-1.2            tibble_3.0.3             
 [85] tsne_0.1-3                crayon_1.3.4              car_3.0-9                
 [88] viridis_0.5.1             jpeg_0.1-8.1              GetoptLong_1.0.2         
 [91] data.table_1.13.0         Rgraphviz_2.32.0          forcats_0.5.0            
 [94] digest_0.6.25             RcppParallel_5.0.2        munsell_0.5.0            
 [97] viridisLite_0.3.0         beeswarm_0.2.3            CytofRUV_0.1             
[100] vipor_0.4.5              

Thanks so much,
Marie

**Comment by mikejiang** [comment]
NA value for `PnR` is not allowed based on FCS standard. You may want to trace back to the software where this faulty file was generated from to see why it happened and get the correct version of the file.

**Comment by mtrussart** [comment]
Hi! Thanks for your quick answer, I have been doing the IMD -> FCS without randomisation on this file from the CyTOF software. How can I read this type of files without randomisation then?
Thanks a lot for your help.

**Comment by SamGG** [comment]
Hi!
Never noticed such a problem. As $P1 is Time, I don't think randomization is applied to it. If you apply randomization the problem disappear? Is it a very short or very long acquisition?
Best.

**Comment by mtrussart** [comment]
Hi!

I send you attached the file as suggested by email:
[A1.fcs.zip](https://github.com/RGLab/flowCore/files/5570319/A1.fcs.zip)

Thanks so much for your help,
Marie

**Comment by SamGG** [comment]
Hi,

Long story made short:
Does this FCS results from a simple IMD to FCS conversion?
Try to update your bioconductor setup in order to read this FCS with a warning instead of an error.

Personal view about the flowCore parameters in FCS:
The flowCore min and max keywords seem to have precedence over the standard parameters. This is leading to troubles sometimes. While my opinion might hurt the guys that are doing a great work in the flowCore ecosystem, I still don't understand why those flowCore parameters are stored in a standardized format that don't required them. Of course, no software need to care about them. But I feel that if I don't take care of them and swap/remove/insert channels and write a compliant FCS, read.FCS() will be lost when reading such a file where the values could exceed their flowCore max or where there are les flowCore min/max than the number of channels.

Long story follows.

Best,
Samuel

With the latest version of flowCore, I only got a warning, and flowCore determines the min and max from the data.

```
> ff = read.FCS("C:/demo/201120-mtrussart/A1.fcs")
Warning messages:
1: In readFCSdata(con, offsets, txt, transformation, which.lines, scale,  :
  $PnR NA is invalid. If it is a numeric string, this could be because it is larger than R's numeric limit: 1.79769313486232e+308. The assigned $PnR value will be imputed from the maximum value of the data in this channel.
2: In FUN(X[[i]], ...) : NAs introduced by coercion
> ff
flowFrame object '190912_RUV_1B_01_cells_found_concat.fcs'
with 319125 cells and 45 observables:
             name            desc       range  minRange    maxRange
$P1          Time              NA 51539607553         0 51539607552
$P2  Event_length              NA         116         0         115
$P3         Y89Di        89Y_CD45        2410         0        2409
$P4       Pd102Di   102Pd_Barcode       20686         0       20685
$P5       Pd104Di   104Pd_Barcode       24756         0       24755
...           ...             ...         ...       ...         ...
$P41      Ir191Di      191Ir_DNA1       31358         0       31357
$P42      Ir193Di      193Ir_DNA2       40098         0       40097
$P43      Pt194Di 194Pt_Cisplatin       44830         0       44829
$P44      Pt195Di 195Pt_Cisplatin       45817         0       45816
$P45      Bi209Di      209Bi_CD61       59192         0       59191
337 keywords are stored in the 'description' slot
> head(ff)
           Time Event_length      Y89Di    Pd102Di    Pd104Di    Pd105Di   Pd106Di    Pd108Di   Pd110Di   In115Di    La139Di
[1,]  573174016           10   0.000000   8.392304   3.729913   3.729913  2.797435  0.9324783  0.000000  0.000000   2.797435
[2,] 1222411008           11   0.000000   4.631971   4.631971   3.087981  2.315985  0.0000000  1.543990  0.000000   0.000000
[3,] 1637049984           15   0.000000   9.586806   6.100695   6.972223  0.000000  1.7430557  3.486111  3.486111  31.375002
[4,] 1820450048           20   2.296351  98.743111 110.990318 100.274017 60.470589 80.3722992 85.730453  2.296351   0.000000
[5,] 2133867008           21  38.876549  17.587009  20.363907  21.289539  7.405057  6.4794250 15.735746 30.545860 126.811600
[6,] 2335657984           19 229.442642 230.375336 234.106110 229.442642 12.125017  2.7980809  0.000000 27.048117   1.865387
        Ce140Di    Pr141Di   Nd142Di    Nd143Di    Nd144Di   Nd145Di   Nd146Di    Sm147Di   Nd148Di   Eu151Di   Eu153Di
[1,]  0.9324783  0.0000000  0.000000  0.0000000  0.0000000  0.000000  0.000000  0.0000000  0.000000  0.000000  1.864957
[2,]  0.0000000  0.0000000  0.000000  0.0000000  0.0000000  0.000000  0.000000  0.0000000  0.000000  0.000000  0.000000
[3,]  0.0000000  4.3576393  0.000000  0.8715279  0.0000000  1.743056  3.486111  7.8437505  0.000000  3.486111  0.000000
[4,]  0.7654505  0.7654505  1.530901  0.0000000  0.7654505  0.000000  1.530901  3.0618019  0.000000  0.000000  0.000000
[5,] 29.6202278 77.7530975 20.363907 18.5126419 12.0332174  5.553792 72.199303 40.7278137 10.181953 50.909767 47.207237
[6,] 13.0577106  5.5961618  0.000000  4.6634684 90.4712830 23.317341  6.528855  0.9326937  6.528855  9.326937  2.798081
        Sm154Di    Gd155Di     Gd156Di    Gd157Di    Tb159Di    Gd160Di  Dy161Di   Dy163Di   Ho165Di    Er166Di  Er167Di
[1,]  0.0000000  0.0000000    1.864957  0.0000000   0.000000  0.0000000 0.000000 0.0000000  0.000000  0.0000000 0.000000
[2,]  0.0000000  0.0000000    0.000000  0.0000000   0.000000  0.0000000 0.000000 0.0000000  0.000000  0.0000000 0.000000
[3,]  0.0000000  1.7430557    0.000000  0.8715279  18.302084  0.8715279 0.000000 0.0000000  0.000000  0.0000000 0.000000
[4,]  0.7654505  0.7654505    0.000000  0.0000000   3.061802  1.5309010 2.296351 0.7654505  1.530901  0.7654505 0.000000
[5,] 33.3227577 49.9841347   17.587009  6.4794250 113.852753 14.8101139 7.405057 2.7768962 45.355972 14.8101139 3.702528
[6,]  0.9326937 17.7211800 1047.414917 29.8461971 120.317482  8.3942432 1.865387 0.0000000  2.798081  0.9326937 1.865387
       Er168Di     Tm169Di    Er170Di   Yb171Di  Yb172Di   Yb173Di   Yb174Di    Ir191Di   Ir193Di     Pt194Di    Pt195Di
[1,] 0.9324783   0.9324783  3.7299132 0.0000000 0.000000 0.0000000 0.0000000   97.91022  196.7529   40.096565   52.21878
[2,] 0.0000000   0.0000000  0.0000000 0.0000000 0.000000 0.0000000 0.0000000   71.79555  128.9232   16.211899   35.51178
[3,] 0.0000000  11.3298616  0.8715279 0.0000000 0.000000 3.4861114 0.0000000  208.29515  392.1875   58.392365   73.20834
[4,] 1.5309010   4.5927029  4.5927029 0.0000000 0.000000 0.0000000 0.7654505 1840.90845 3507.2942 1096.890503 1062.44531
[5,] 4.6281605 174.0188446 41.6534462 1.8512642 1.851264 0.9256321 3.7025285  211.04413  415.6088  174.944473  166.61378
[6,] 2.7980809  13.0577106  9.3269367 0.9326937 1.865387 0.0000000 1.8653873 1076.32849 2053.7915    9.326937    0.00000
         Bi209Di
[1,]   0.0000000
[2,]   0.7719951
[3,]   6.1006947
[4,]   0.7654505
[5,]  21.2895393
[6,] 533.5007935
> packageVersion("flowCore")
[1] ‘2.2.0’
```

The time values are excessively large, and larger than the number bits declared. While time is displayed as an integer value in R, I think time is encoded as a float as all the data in most of FCS nowadays. The $DATATYPE is pointing that. All acquisition time points are ordered increasingly, which is fine.
```
> keyword(ff, "$DATATYPE")
$`$DATATYPE`
[1] "F"
> all(diff(ff_time) >= 0)
[1] TRUE
```

Plotting any marker and the time shows the time increasing and a correlation of time shifts with shifts of markers (here it is a barcoding marker).
```
y1 = asinh(exprs(ff)[,5]/5)
y2 = ff_time / max(ff_time)
plot(y1, pch = '.')
points(y2 * max(y1)/3, pch = '.', col = "red")
```
![image](https://user-images.githubusercontent.com/3333348/99840273-87d0c300-2b6c-11eb-96ef-62b2e9cfa4e2.png)

Back to the time parameter, this is the flowCore max parameter that is set to NA. I just can imagine that it is set to NA because this maximum exceeds the highest integer value (2^31 or 2^32). There is no NA in the time channel and all values are ordered increasingly.
```
> all(diff(ff_time) >= 0)
[1] TRUE
> any(is.na(ff_time))
[1] FALSE
> (idx = grep("\\$P1\\D", names(keyword(ff)), perl = TRUE, value = TRUE))
[1] "$P1B"             "$P1E"             "$P1N"             "$P1R"             "flowCore_$P1Rmax" "flowCore_$P1Rmin"
> keyword(ff, idx)
$`$P1B`
[1] "32"
$`$P1E`
[1] "0,0"
$`$P1N`
[1] "Time"
$`$P1R`
[1] 51539607553
$`flowCore_$P1Rmax`
[1] "NA"
$`flowCore_$P1Rmin`
[1] "0"
```

As far as I understand the FCS standard, as the datatype is F, ie single precision FLOAT, the maximum should be the highest floating value encoded in 32 bits, which should be around 1e38.

Writing the flowFrame and reading back leads to the following output. I don't understand the NA by coercion warning.
```
> write.FCS(ff, "C:/demo/201120-mtrussart/A1_b.fcs")
[1] "C:/demo/201120-mtrussart/A1_b.fcs"
> 
> (ff2 = read.FCS("C:/demo/201120-mtrussart/A1_b.fcs"))
flowFrame object '190912_RUV_1B_01_cells_found_concat.fcs'
with 319125 cells and 45 observables:
             name            desc       range  minRange    maxRange
$P1          Time              NA 51539607554         0 51539607553
$P2  Event_length              NA         116         0         115
$P3         Y89Di        89Y_CD45        2410         0        2409
$P4       Pd102Di   102Pd_Barcode       20686         0       20685
$P5       Pd104Di   104Pd_Barcode       24756         0       24755
...           ...             ...         ...       ...         ...
$P41      Ir191Di      191Ir_DNA1       31358         0       31357
$P42      Ir193Di      193Ir_DNA2       40098         0       40097
$P43      Pt194Di 194Pt_Cisplatin       44830         0       44829
$P44      Pt195Di 195Pt_Cisplatin       45817         0       45816
$P45      Bi209Di      209Bi_CD61       59192         0       59191
337 keywords are stored in the 'description' slot
Warning message:
In FUN(X[[i]], ...) : NAs introduced by coercion
```

Finally finding 'concat' in the file name and viewing the previous plot, I doubt that this file results from a simple IMD to FCS conversion. Let us know.

**Comment by mtrussart** [comment]
Thanks so much for looking into the file and sharing your thoughts on why those NA values are appearing. This files has been processed by CATALYST for debarcoding, when I went up to understand the error I went up to the flowCore functions and I found out that the problem was coming from the $P1R that has been saved as NA. And I did not encounter this problem when using files without changing this step in the IMD conversion.

Thanks a lot for all your help with this,
Best wishes,
Marie
